lib/similarity_utils.py

```diff
@@ -21,3 +22,70 @@ def iou(lista1: List[Any], lista2: List[Any]) -> float:
         intersection = set(lista1).intersection(set(lista2))
         union = set(lista1).union(set(lista2))
         return len(intersection) / len(union)
+
+
+def sort_two_lists(list1: List[Any], list2: List[Any], reverse: bool = True):
+    """metoda sortuje dwie listy równocześnie
+    list1 jest sortowana po wartościach, a kolejność elementów w list2 zależy od sortowania list1
+    reverse = True oznacza kolejność malejącą, a reverse = False oznacza kolejność rosnącą"""
+    zipped_lists = zip(list1, list2)
+    list1, list2 = zip(*sorted(zipped_lists, reverse=reverse))
+    return list1, list2
+
+
+def find_actor(data: pd.DataFrame, actor_id: str) -> List[Any]:
+    """metoda na podstawie ramki danych oraz id aktora, zwraca jego dane z ramki danych"""
+    actor_index = data[data['nconst'] == actor_id].index
+    return data.iloc[actor_index[0]]
+
+
+def prepare_pandas_row(pandas_row: pd.DataFrame) -> List[Any]:
+    """metoda zamienia wiersz wydobyty z Pandas DataFrame i przekształca go w pythonową listę"""
+    p_list = []
+    for value in pandas_row:
+        try:
+            p_list.append(value.tolist())
+        except AttributeError:
+            p_list.append(value)
+    return p_list
+
+
+def similarity(actor1: List[Any], actor2: List[Any]) -> float:
+    """metoda licząca similarity pomiędzy dwoma aktorami
+    jej argumentami są dwie listy, a wartością wyjściową wartość z przedziału [-1, 1]
+    metoda jest przygotowana pod dane ze zbioru JOINED_DATA"""
+    weights = [0.3, 0.2, 0.3, 0.2]
+    values = [
+        iou(actor1[1], actor2[1]),           # similarty ze względu na ilość wspólnych filmów
+        iou(actor1[2], actor2[2]),           # similarity ze względu na rodzaj granych produkcji
+        iou(actor1[7], actor2[7]),           # similarity ze względu na gatunek granych produkcji
+        1 if actor1[8] == actor2[8] else 0   # similarity ze względu na tę samą płeć
+    ]
+    #TODO dodać linijkę uwzględniającą kolumnę knownForTitles za pomocą metody iou
+    return sum(weights[i] * values[i] for i in range(4)) * 2 - 1
+
+
+def similarity_one_vs_all(data: pd.DataFrame, main_actor: List[Any]):
+    """metoda liczy similarity pomiędzy aktorem main_actor, a wszystkimi aktorami obecnymi w ramce danych data
+    każdy wiersz ramki jest zamieniany na listę, a nastepnie do uzsykanej listy i main_actor przykładana jest funkcja similarity"""
+    ids = []
+    similarities = []
+    for i in range(len(data)):
+        actor = prepare_pandas_row(data.loc[i].values.tolist())
+        ids.append(actor[0])
+        similarities.append(similarity(main_actor, actor[0:]))
+    return ids, similarities
+
+
+def print_most_similiar_actors(data: pd.DataFrame, main_actor: List[Any], ids: List[str], values: List[float],
```

Comment:
brakuje mi jeszcze osobnej funkcji zwracającej ranking podobnych aktorów. Jest ona potrzebna do metryk.

Reply:
Dzisiaj w nocy spróbuję to dodać, chociaż to chyba trochę wychodzi poza scope zadania

Reply:
W sumie to nie mamy jeszcze metryki, więc proponuję to rozwiązać poprzez dodanie zadania na click-upie. Wykracza to trochę poza "liczenie similarity"

Reply:
Za tamto pierwsze z sprintu 1 pewnie tak, ale skoro zamykamy je w sprintcie 2 to chciałbym to mieć.  
Chodzi mi jedynie aby dołożyć jeden dodatkowy interfejs we wnętrzu którego:
wywołujesz similarity_one_vs_all
sortujesz id-ki względem tego similarity
zwracasz określoną czołówkę tego rankingu np `top: int = 5`

Reply:
Dodałem metodę z rankingowaniem